applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_gradient_projection.py

```diff
@@ -184,71 +148,103 @@ def __analyzeShape(self):
-        self.optimization_utilities.CalculateProjectedSearchDirectionAndCorrection(
+        optimization_utilities.CalculateProjectedSearchDirectionAndCorrection(
             nabla_f,
             N,
             g_a,
             solver,
             s,
             c)
+        eps = 1e-16
 
-        if c.norm_inf() != 0.0:
+        if c.norm_inf() > eps:
             if c.norm_inf() <= self.max_correction_share * self.step_size:
                 delta = self.step_size - c.norm_inf()
-                s *= delta/s.norm_inf()
+                s_inf_norm = s.norm_inf()
+                if s_inf_norm > eps :
+                    s *= delta/s_inf_norm
+                else:
+                    s *= delta
             else:
                 KM.Logger.PrintWarning("ShapeOpt", f"Correction is scaled down from {c.norm_inf()} to {self.max_correction_share * self.step_size}.")
-                c *= self.max_correction_share * self.step_size / c.norm_inf()
-                s *= (1.0 - self.max_correction_share) * self.step_size / s.norm_inf()
+                c_inf_norm = c.norm_inf()
+                s_inf_norm = s.norm_inf()
+                if c_inf_norm > eps:
+                    c *= self.max_correction_share * self.step_size / c_inf_norm
+                else:
+                    c *= self.max_correction_share * self.step_size
+                if s_inf_norm > eps:
+                    s *= (1.0 - self.max_correction_share) * self.step_size / s_inf_norm
+                else:
+                    s *= (1.0 - self.max_correction_share) * self.step_size
         else:
-            s *= self.step_size / s.norm_inf()
+            s_inf_norm = s.norm_inf()
+            if s_inf_norm > eps:
+                s *= self.step_size / s_inf_norm
+            else :
+                s *= self.step_size
+
+        s_plus_c = KM.Vector(s+c)
+
 
-        self.optimization_utilities.AssignVectorToVariable(self.design_surface, s, KSO.SEARCH_DIRECTION)
-        self.optimization_utilities.AssignVectorToVariable(self.design_surface, c, KSO.CORRECTION)
-        self.optimization_utilities.AssignVectorToVariable(self.design_surface, s+c, KSO.CONTROL_POINT_UPDATE)
+        previous_end = 0
+        for design_surface in self.design_surfaces.values():
+            initial_index = previous_end
+            final_index = initial_index + design_surface.NumberOfNodes()*3
+            previous_end = final_index
+
+            optimization_utilities.AssignVectorToVariable(design_surface, self.__GetVector(s[initial_index:final_index]), KSO.SEARCH_DIRECTION)
+            optimization_utilities.AssignVectorToVariable(design_surface, self.__GetVector(c[initial_index:final_index]), KSO.CORRECTION)
+            optimization_utilities.AssignVectorToVariable(design_surface, self.__GetVector(s_plus_c[initial_index:final_index]), KSO.CONTROL_POINT_UPDATE)
+
+    def __GetVector(self, vec_slice):
+        return KM.Vector([i for i in vec_slice])
+
+        return KM.Vector(vec_list)
```

Comment:
```suggestion
    def __GetVector(self, vec_slice):
        return KM.Vector(vec_slice)
```

I would say that this should work too ...
Can you try?

Reply:
I tried and it did not work. as vec_slice is type `slice`  and not a vector

Reply:
you mean not a `list`
can you post the error?

I am curious as maybe we can add it with little effort, see here:
https://github.com/KratosMultiphysics/Kratos/blob/c66a0d83b732054cec11fcbf251057efa359179c/kratos/python/add_vector_to_python.cpp#L213-L218

Reply:
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. Kratos.Vector()
    2. Kratos.Vector(arg0: Kratos.Vector)
    3. Kratos.Vector(arg0: int)
    4. Kratos.Vector(arg0: int, arg1: float)
    5. Kratos.Vector(arg0: Kratos.Vector)
    6. Kratos.Vector(arg0: Kratos::array_1d<double, 3ul>)
    7. Kratos.Vector(arg0: list)

Invoked with: <Kratos.VectorSlice object at 0x7f987b362ab0>

Reply:
thx

@ashishdarekar this could be a task for you, I don't think it is difficult to do
Lets speak about this in our next meeting

Reply:
Its just that that constructor is not exposed and it should be done. I did not do it here as it should not be a part of this PR !

Reply:
Resolved in PR #8896 
Now you can use this `return KM.Vector(vec_slice)`